### PR TITLE
Figure.ternary: Support Frame's xaxis/yaxis/zaxis attributes to set axis settings for a/b/c axes

### PR DESCRIPTION
**Pull request by Chuan1937**
Add ternary frame 

Update the test `pygmt/tests/test_ternary.py` to use the new implemented Frame and Axis parameter classes.

Fixes partly https://github.com/GenericMappingTools/pygmt/issues/4494

### CODE REVIEW

**Comment by seisman** [comment]
xref: https://github.com/GenericMappingTools/pygmt/pull/4505#discussion_r3051550813

**Comment by yvonnefroehlich** [comment]
Actually I would split this into two PRs, one for the code changes for the ternary frame itself and one for updating the tests.

File: pygmt/src/ternary.py [hunk 1]
```diff
@@ -104,7 +104,22 @@ def ternary(  # noqa: PLR0913
 
     with Session() as lib:
         with lib.virtualfile_in(check_kind="vector", data=data) as vintbl:
+            kwargs = build_arg_list(aliasdict, infile=vintbl)
```
Comment (seisman) [inline pygmt/src/ternary.py]: Instead of looping over the argument list, I think we should check the `frame` parameter directly, something like:

```
if isinstance(frame, (Frame, Axis)):
   ...
```

Also need to note that, for `ternary`, `axes` setting like `WSEN` is not allowed and there are no primary/secondary axis. This can be verified by:
```
gmt begin map
  gmt makecpt -Cturbo -T0/80/10
  gmt ternary @ternary.txt -R0/100/0/100/0/100 -JX15c -Sc0.1c -C -LLimestone/Water/Air -BWSen+tABC -Bafg
gmt end show
```
and 
```
gmt begin map
  gmt makecpt -Cturbo -T0/80/10
  gmt ternary @ternary.txt -R0/100/0/100/0/100 -JX15c -Sc0.1c -C -LLimestone/Water/Air -B+tABC -Bpaafg -Bpbafg -Bpcafg
gmt end show
```

**Comment by seisman** [comment]
> Actually I would split this into two PRs, one for the code changes for the ternary frame itself and one for updating the tests.

Actually we have to update the tests in this PR to ensure the code changes are correct.

File: pygmt/src/ternary.py [hunk 2]
```diff
@@ -90,6 +92,43 @@ def ternary(  # noqa: PLR0913
     _labels = [v if v is not None else "-" for v in (alabel, blabel, clabel)]
     labels = _labels if any(v != "-" for v in _labels) else None
 
+    # Convert Frame/Axis to ternary-compatible format.
```
Comment (seisman) [inline pygmt/src/ternary.py]: Please put it in a private function, then we can test the logic directly in doctests.

File: pygmt/tests/test_ternary.py [hunk 3]
```diff
@@ -99,8 +109,58 @@ def test_ternary_1_label(array):
...
+    """
+    fig = Figure()
+    with pytest.raises(GMTParameterError, match="secondary axes"):
+        fig.ternary(
+            data=array,
+            region=[0, 100, 0, 100, 0, 100],
+            width="10c",
+            frame=Frame(xaxis2=Axis(annot=True)),
+            style="c0.1c",
+        )
```
Comment (seisman) [inline pygmt/tests/test_ternary.py]: These two tests can be removed because they're already covered in the doctests of the private function.

File: pygmt/src/ternary.py [hunk 4]
```diff
@@ -8,7 +8,95 @@
...
+            parts.append(str(frame.axis))
+        # Per-axis: xaxis→a, yaxis→b, zaxis→c
+        for ternary_prefix, axis_obj in [
+            ("a", frame.xaxis),
+            ("b", frame.yaxis),
+            ("c", frame.zaxis),
+        ]:
+            if axis_obj:
+                parts.append(f"{ternary_prefix}{axis_obj}")
+        return parts
```
Comment (seisman) [inline pygmt/src/ternary.py]: It seems the current code only support `title` but not `fill` or other frame settings. I think we should have a more general way to handle it.

File: pygmt/src/ternary.py [hunk 5]
```diff
@@ -2,13 +2,112 @@
...
+                reason="For ternary diagrams, Frame.axes (e.g., 'WSen') is not supported.",
+            )
+        if any((frame.xaxis2, frame.yaxis2, frame.zaxis2)):
+            raise GMTParameterError(
+                conflicts_with=(
+                    "frame",
+                    ["frame.xaxis2", "frame.yaxis2", "frame.zaxis2"],
+                ),
+                reason="For ternary diagrams, secondary axes are not supported.",
+            )
```
Comment (seisman) [inline pygmt/src/ternary.py]: I've checked the `-B` option. It turns out the in ternary, only a few modifiers are supported. So, a whitelist of allowed attributes makes more sense and is easier to maintain.
```suggestion
        _attributs = ["title", "subtitle", "fill", "axis", "xaxis", "yaxis", "zaxis"]
        if any(
            getattr(frame, attr) and attr not in _attributs for attr in vars(frame)
        ):
            raise GMTValueError(
                repr(frame),
                description="frame setting",
                reason="For ternary diagrams, only Frame attributes " + f"{', '.join(repr(_attr) for _attr in _attributs)} are supported."
            )
```

File: pygmt/src/ternary.py [hunk 6]
```diff
@@ -2,13 +2,112 @@
...
+            parts.append(str(frame.axis))
+        # Per-axis: xaxis→a, yaxis→b, zaxis→c
+        for ternary_prefix, axis_obj in [
+            ("a", frame.xaxis),
+            ("b", frame.yaxis),
+            ("c", frame.zaxis),
+        ]:
+            if axis_obj:
+                parts.append(f"{ternary_prefix}{axis_obj}")
+        return parts
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
        frame_settings = _Axes(title=frame.title, subtitle=frame.subtitle, fill=frame.fill)
        return [
            Alias(frame_settings) if str(frame_settings) else Alias(None),
            Alias(frame.axis),
            Alias(frame.xaxis, prefix="a"),
            Alias(frame.yaxis, prefix="b"),
            Alias(frame.zaxis, prefix="c"),
        ]
```

File: pygmt/src/ternary.py [hunk 7]
```diff
@@ -8,7 +8,94 @@
...
+        frame_settings = _Axes(
+            title=frame.title, subtitle=frame.subtitle, fill=frame.fill
+        )
+        return [
+            Alias(frame_settings) if str(frame_settings) else Alias(None),
+            Alias(frame.axis),
+            Alias(frame.xaxis, prefix="a"),
+            Alias(frame.yaxis, prefix="b"),
+            Alias(frame.zaxis, prefix="c"),
+        ]
```
Comment (seisman) [inline pygmt/src/ternary.py]: Changes in lines 179-192 are not needed if applying changes below:
```suggestion
        paras = [
            Alias(frame_settings) if str(frame_settings) else Alias(None),
            Alias(frame.axis),
            Alias(frame.xaxis, prefix="a"),
            Alias(frame.yaxis, prefix="b"),
            Alias(frame.zaxis, prefix="c"),
        ]
        return [par._value for par in paras if par._value is not None]
```

File: pygmt/src/ternary.py [hunk 8]
```diff
@@ -8,7 +8,92 @@
...
+        The converted frame parameter. For Frame inputs, returns a list of strings;
+        for Axis, str, bool, or list inputs, returns the value directly.
+
+    Examples
+    --------
+    >>> from pygmt.params import Axis, Frame
+    >>> _ternary_frame(Axis(annot=True, tick=True, grid=True))
+    'afg'
+    >>> _ternary_frame(Axis(annot=True, tick=True))
+    'af'
```
Comment (seisman) [inline pygmt/src/ternary.py]: A little duplicate with the previous one.
```suggestion
```

File: pygmt/src/ternary.py [hunk 9]
```diff
@@ -8,7 +8,92 @@
...
+    ... )
+    ['+tTitle', 'aafg+lWater', 'bafg+lAir', 'cafg+lLimestone']
+    >>> _ternary_frame(Frame(fill="lightblue", axis=Axis(annot=True)))
+    ['+glightblue', 'a']
+    >>> _ternary_frame("afg")
+    'afg'
+    >>> _ternary_frame(True)
+    True
+    >>> _ternary_frame(["afg", "aafg+lWater"])
+    ['afg', 'aafg+lWater']
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
    >>> _ternary_frame(["aafg+lWater", "bafg+lAir", "cafg+lLimestone"])
    ['aafg+lWater', 'bafg+lAir', 'cafg+lLimestone']
```

File: pygmt/src/ternary.py [hunk 10]
```diff
@@ -8,7 +8,92 @@
...
+    >>> _ternary_frame(Frame(axes="WSen", axis=Axis(annot=True)))
```
Comment (seisman) [inline pygmt/src/ternary.py:65]:
```suggestion
    >>> _ternary_frame("none")
    'none'
    >>> _ternary_frame(Frame(axes="WSen", axis=Axis(annot=True)))
```

File: pygmt/src/ternary.py [hunk 11]
```diff
@@ -85,15 +170,18 @@ def ternary(  # noqa: PLR0913
...
     """
     self._activate_figure()
-
     # -Lalabel/blabel/clabel. '-' means skipping the label.
     _labels = [v if v is not None else "-" for v in (alabel, blabel, clabel)]
     labels = _labels if any(v != "-" for v in _labels) else None
 
+    # Convert Frame/Axis to ternary-compatible format.
+    frame_option = _ternary_frame(frame)
+    option_b = Alias("+n" if frame_option == "none" else frame_option, name="frame")
```
Comment (seisman) [inline pygmt/src/ternary.py]: I don't think we need this line. Just pass `B=_ternary_frame(frame)` should work. 

BTW, we also need to add a test for `frame="none"`.

File: pygmt/src/ternary.py [hunk 12]
```diff
@@ -8,7 +8,92 @@
...
 from pygmt.clib import Session
+from pygmt.exceptions import GMTValueError
 from pygmt.helpers import build_arg_list, fmt_docstring, use_alias
+from pygmt.params import Axis, Frame
+from pygmt.params.frame import _Axes
+
+
+def _ternary_frame(frame):
+    """
+    Convert a Frame/Axis parameter to ternary-compatible format.
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
    Convert 'frame' to ternary-compatible format.
```

File: pygmt/src/ternary.py [hunk 13]
```diff
@@ -8,7 +8,92 @@
...
+
+    For ternary diagrams, GMT uses axis names **a**, **b**, **c** instead of
+    **x**, **y**, **z**, and there are no primary/secondary axes. This function
+    converts a :class:`pygmt.params.Frame` or :class:`pygmt.params.Axis` object
+    to a list of strings with the correct ternary prefixes.
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
    to a string or a list of strings with the correct axis prefixes.
```

File: pygmt/src/ternary.py [hunk 14]
```diff
@@ -8,7 +8,92 @@
...
+    >>> _ternary_frame(Frame(xaxis2=Axis(annot=True)))
+    Traceback (most recent call last):
+    pygmt.exceptions.GMTValueError: ...
+    """
+    if isinstance(frame, Axis):
+        return str(frame)
+    if isinstance(frame, Frame):
+        _attributes = ["title", "subtitle", "fill", "axis", "xaxis", "yaxis", "zaxis"]
+        if any(
+            getattr(frame, attr) and attr not in _attributes for attr in vars(frame)
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
            _attr not in _attributes and getattr(frame, _attr) for _attr in vars(frame)
```

File: pygmt/src/ternary.py [hunk 15]
```diff
@@ -8,7 +8,92 @@
...
+            raise GMTValueError(
+                repr(frame),
+                description="frame setting",
+                reason="For ternary diagrams, only Frame attributes "
+                f"{', '.join(repr(_attr) for _attr in _attributes)} are supported.",
+            )
+        frame_settings = _Axes(
+            title=frame.title, subtitle=frame.subtitle, fill=frame.fill
+        )
+        paras = [
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
        params = [
```

File: pygmt/src/ternary.py [hunk 16]
```diff
@@ -8,7 +8,92 @@
...
+            Alias(frame_settings) if str(frame_settings) else Alias(None),
+            Alias(frame.axis),
+            Alias(frame.xaxis, prefix="a"),
+            Alias(frame.yaxis, prefix="b"),
+            Alias(frame.zaxis, prefix="c"),
+        ]
+        return [par._value for par in paras if par._value is not None]
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
        return [par._value for par in params if par._value is not None]
```

File: pygmt/src/ternary.py [hunk 17]
```diff
@@ -85,15 +170,18 @@ def ternary(  # noqa: PLR0913
...
     # -Lalabel/blabel/clabel. '-' means skipping the label.
     _labels = [v if v is not None else "-" for v in (alabel, blabel, clabel)]
     labels = _labels if any(v != "-" for v in _labels) else None
 
+    # Convert Frame/Axis to ternary-compatible format.
+    frame_option = _ternary_frame(frame)
+    option_b = Alias("+n" if frame_option == "none" else frame_option, name="frame")
+
     aliasdict = AliasSystem(
+        B=option_b,
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
```

File: pygmt/src/ternary.py [hunk 18]
```diff
@@ -85,15 +170,18 @@ def ternary(  # noqa: PLR0913
...
     self._activate_figure()
-
     # -Lalabel/blabel/clabel. '-' means skipping the label.
     _labels = [v if v is not None else "-" for v in (alabel, blabel, clabel)]
     labels = _labels if any(v != "-" for v in _labels) else None
 
+    # Convert Frame/Axis to ternary-compatible format.
+    frame_option = _ternary_frame(frame)
+    option_b = Alias("+n" if frame_option == "none" else frame_option, name="frame")
+
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
```

File: pygmt/src/ternary.py [hunk 19]
```diff
@@ -85,15 +170,18 @@ def ternary(  # noqa: PLR0913
...
     aliasdict = AliasSystem(
+        B=option_b,
         L=Alias(labels, name="alabel/blabel/clabel", sep="/", size=3),
     ).add_common(
-        B=frame,
         R=region,
```
Comment (seisman) [inline pygmt/src/ternary.py:185]:
```suggestion
        B=_ternary_frame(frame),
        R=region,
```

File: pygmt/src/ternary.py [hunk 20]
```diff
@@ -8,7 +8,92 @@
...
+
+
+def _ternary_frame(frame):
+    """
+    Convert 'frame' to ternary-compatible format.
+
+    For ternary diagrams, GMT uses axis names **a**, **b**, **c** instead of
+    **x**, **y**, **z**, and there are no primary/secondary axes. This function
+    converts a :class:`pygmt.params.Frame` or :class:`pygmt.params.Axis` object
+    to a string or a list of strings with the correct axis prefixes.
```
Comment (seisman) [inline pygmt/src/ternary.py]: Please rewrap these line to 88 characters.

File: pygmt/src/ternary.py [hunk 21]
```diff
@@ -20,7 +105,7 @@ def ternary(  # noqa: PLR0913
     blabel: str | None = None,
     clabel: str | None = None,
     region: Sequence[float | str] | str | None = None,
-    frame: str | Sequence[str] | Literal["none"] | bool = False,
+    frame: str | Sequence[str] | Literal["none"] | bool | Frame | Axis = False,
     verbose: Literal["quiet", "error", "warning", "timing", "info", "compat", "debug"]
     | bool = False,
     panel: int | Sequence[int] | bool = False,
```
Comment (seisman) [inline pygmt/src/ternary.py:111]: It seems we don't have `$frame` in the docstrings. Please add it and clarify that use `Frame`'s xaxis/yaxis/zaxis attributes to set the a/b/c axes of ternary diagrams.

File: pygmt/tests/test_ternary.py [hunk 22]
```diff
@@ -99,8 +108,28 @@ def test_ternary_1_label(array):
...
+            yaxis=Axis(annot=True, tick=True, grid=True, label="Air"),
+            zaxis=Axis(annot=True, tick=True, grid=True, label="Limestone"),
+        ),
         style="c0.1c",
         pen="thinnest",
     )
     return fig
+
+
+def test_ternary_axis(array):
```
Comment (seisman) [inline pygmt/tests/test_ternary.py:122]: Please add a new test to ensure `frame="none"` works as expected.

File: pygmt/tests/test_ternary.py [hunk 23]
```diff
@@ -99,8 +108,44 @@ def test_ternary_1_label(array):
...
+            xaxis=Axis(annot=True, tick=True, grid=True, label="Water"),
+            yaxis=Axis(annot=True, tick=True, grid=True, label="Air"),
+            zaxis=Axis(annot=True, tick=True, grid=True, label="Limestone"),
+        ),
         style="c0.1c",
         pen="thinnest",
     )
     return fig
+
+
```
Comment (seisman) [inline pygmt/tests/test_ternary.py:121]: Need to add the `@pytest.mark.mpl_image_compare` marker and add the baseline images.
```suggestion
@pytest.mark.mpl_image_compare
```

File: pygmt/tests/test_ternary.py [hunk 24]
```diff
@@ -99,8 +108,44 @@ def test_ternary_1_label(array):
...
+        data=array,
+        region=[0, 100, 0, 100, 0, 100],
+        cmap="red,orange,yellow,green,blue,violet",
+        width="10c",
+        frame=Axis(annot=True, tick=True, grid=True),
+        style="c0.1c",
+        pen="thinnest",
+    )
+
+
```
Comment (seisman) [inline pygmt/tests/test_ternary.py]:
```suggestion
@pytest.mark.mpl_image_compare
```

**Comment by seisman** [comment]
Still need to generate the baseline images and put them to dagshug, following the instructions at https://www.pygmt.org/dev/contributing.html#using-data-version-control-dvc-to-manage-test-images. Let me know if you'd like to do it or need help.

**Comment by Chuan1937** [comment]
> Still need to generate the baseline images and put them to dagshug, following the instructions at https://www.pygmt.org/dev/contributing.html#using-data-version-control-dvc-to-manage-test-images. Let me know if you'd like to do it or need help.

ok, I can do it.

**Comment by Chuan1937** [comment]
@seisman
I tried pushing to DagsHub, but I'm not a collaborator yet.

**Comment by seisman** [comment]
You need to create an account on DagsHub, and then I can add you as a collaborator

**Comment by Chuan1937** [comment]
> You need to create an account on DagsHub, and then I can add you as a collaborator

I have logged in using github.

**Comment by seisman** [comment]
Hmmm. I can't find you https://dagshub.com/Chuan1937

**Comment by Chuan1937** [comment]
> Hmmm. I can't find you https://dagshub.com/Chuan1937

I changed to https://dagshub.com/Chuan1937

**Comment by seisman** [comment]
Ok. DOne

File: pygmt/src/ternary.py [hunk 25]
```diff
@@ -65,6 +150,10 @@ def ternary(  # noqa: PLR0913
         [*amin*, *amax*, *bmin*, *bmax*, *cmin*, *cmax*].
         Give the min and max limits for each of the three axes **a**, **b**,
         and **c**.
+    $frame
+        For ternary diagrams, use :class:`pygmt.params.Frame` ``xaxis``, ``yaxis``,
+        and ``zaxis`` attributes to set the **a**, **b**, and **c** axes,
+        respectively.
```
Comment (seisman) [inline pygmt/src/ternary.py]:
```suggestion
        For ternary diagrams, use :class:`pygmt.params.Frame` ``xaxis``, ``yaxis``, and
        ``zaxis`` attributes to set the **a**, **b**, and **c** axes, respectively.
```

File: pygmt/tests/baseline/test_ternary.png.dvc [hunk 26]
```diff
@@ -1,5 +1,5 @@
 outs:
-- md5: 065ea10a2ed961df29defd6a1dfd1a4c
+- md5: cd47a3947f807e53c30b1efa19202722
```
Comment (seisman) [inline pygmt/tests/baseline/test_ternary.png.dvc]: This baseline doesn't change. So no need to update it.

File: pygmt/tests/baseline/test_ternary_1_label.png.dvc [hunk 27]
```diff
@@ -1,5 +1,5 @@
 outs:
-- md5: 8046b09ffd4020ac84b43704b39597d3
```
Comment (seisman) [inline pygmt/tests/baseline/test_ternary_1_label.png.dvc:2]: Same here.

File: pygmt/tests/baseline/test_ternary_3_labels.png.dvc [hunk 28]
```diff
@@ -1,5 +1,5 @@
 outs:
-- md5: af08eb16a966e9e8b72137c22c880c60
+- md5: 4271affa27340c69bb52659be29c652c
```
Comment (seisman) [inline pygmt/tests/baseline/test_ternary_3_labels.png.dvc]: Same here.

**Comment by seisman** [comment]
I just found one minor issue, which can be reproduced below
```
gmt begin map
  gmt makecpt -Cturbo -T0/80/10
  gmt ternary @ternary.txt -R0/100/0/100/0/100 -JX15c -Sc0.1c -C -Bafg
  gmt ternary @ternary.txt -R0/100/0/100/0/100 -JX15c -Sc0.1c -C -Bafg -B -Xw+1cba
gmt end show
```
map

As you can see, passing `-Bafg` alone doesn't plot the frame, which makes little sense. I think in this case, we should always pass `-Bafg -B` instead.

File: pygmt/tests/test_ternary.py [hunk 29]
```diff
@@ -99,7 +108,65 @@ def test_ternary_1_label(array):
...
+    fig.ternary(
+        data=array,
+        region=[0, 100, 0, 100, 0, 100],
+        cmap="red,orange,yellow,green,blue,violet",
+        width="10c",
+        frame="afg",
+        style="c0.1c",
+        pen="thinnest",
+    )
+    return fig
```
Comment (seisman) [inline pygmt/tests/test_ternary.py]: This test generates the same baseline image as `test_ternary_axis`, so we should use `pytest.parameterize` for combine these tests.

In other words, we should test that, `frame="afg"`, `frame=Axis(annot=True, tick=True, grid=True)`, `frame=Frame(axis=Axis(annot=True, tick=True, grid=True))` generate the same output.

**Review by seisman — APPROVED** [review]
